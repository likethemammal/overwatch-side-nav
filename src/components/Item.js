import React, { Component } from 'react'

import css from './Item.less'

import general from '@likethemammal/general-components'

const { SVG } = general.components

class Item extends Component {
    static defaultProps = {
        className: css.container,
        defaultColor: '#323c50',
        textColor: 'white',
        activeColor: '#00c3ff'
    }

    state = {
        hovered: false,
    }

    onClick = () => {
        const { id, onHashChange } = this.props

        onHashChange(`#${id}`)
    }

    onMouseEnter = () => {
        this.setState({
            hovered: true,
        })
    }

    onMouseLeave = () => {
        this.setState({
            hovered: false,
        })
    }

    render() {

        const {
            className,
            id,
            label,
            icon,
            hash,
            defaultColor,
            textColor,
            activeColor,
        } = this.props

        const {
            hovered
        } = this.state

        const active = hash === `#${id}`
        const colorStyle = {
            backgroundColor: active || hovered ? activeColor : defaultColor,
            color: textColor,
        }

        const iconStyle = {
            fill: textColor,
        }

        return <a
            href={`/#${id}`}
            className={className}
            onClick={this.onClick}
        >
            <div
                style={colorStyle}
                className={css.inner}
                onMouseEnter={this.onMouseEnter}
                onMouseLeave={this.onMouseLeave}
            >
                <div className={css.icon_container}>
                    <SVG
                        raw={icon}
                        className={css.icon}
                        style={iconStyle}
                    />
                </div>
                <div className={css.label}>
                    {label}
                </div>
            </div>
        </a>
    }
}

export default Item