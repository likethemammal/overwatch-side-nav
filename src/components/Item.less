@import "../colors.less";

@item_height: 4.0rem;
@icon_width: 4.3rem;
@hover_padding_size: 1rem;

.container {
    composes: cf from '@likethemammal/general-components/bin/general.css';
    position: relative;
    right: -100%;
    display: block;
    margin-left: -@icon_width - @hover_padding_size;
    padding-left: @hover_padding_size;
    padding-bottom: @hover_padding_size - 0.7rem;
    color: inherit;

    transition: right 275ms ease;

    &:hover {
        right: 0;
    }

    &, & * {
        box-sizing: border-box;
    }
}

.inner {
    text-align: center;
    border-radius: 5px 0 0 5px;

    transition: 100ms linear;
}

.icon_container {
    float: left;
}

.icon {
    height: @item_height;
    width: @icon_width;
    padding: 0 0.8rem;
    border-radius: 3px 0 0 3px;
    background: rgba(0,0,0,0.1);
    composes: container from '@likethemammal/general-components/bin/components/SVG.css';
}

.label {
    display: inline-block;
    line-height: @item_height;
    padding-right: 13px;
    padding-left: 13px;

    font-size: 2rem;
    text-transform: uppercase;
    text-align: center;
    letter-spacing: 0.06rem;
    font-family: 'Futura No 2 D DemiBold', Arial;
}