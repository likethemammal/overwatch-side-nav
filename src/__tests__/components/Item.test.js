import React from 'react';
import {shallow} from 'enzyme';

import general from '@likethemammal/general-components'

const shared = general.__tests__._shared.tests

import Item from '../../components/Item'

const mock_onHashChange = jest.fn()


describe('Item', () => {

    const component = shallow(
        <Item
            icon={'text-file-stub'}
            id={'item_id'}
            label={'item_label'}
            onHashChange={mock_onHashChange}
        />
    )

    shared.SNAPSHOTS_SHOULD_MATCH([
        component
    ])

    //handle hovered state changes

})