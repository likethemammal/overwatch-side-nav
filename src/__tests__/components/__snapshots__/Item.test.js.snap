// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item Snapshot 0 should match 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Item
    activeColor="#00c3ff"
    className={undefined}
    defaultColor="#323c50"
    icon="text-file-stub"
    id="item_id"
    label="item_label"
    onHashChange={[Function]}
    textColor="white"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className={undefined}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "backgroundColor": "#323c50",
            "color": "white",
          }
        }
      >
        <div
          className={undefined}
        >
          <t
            className="SVG-container"
            raw="text-file-stub"
            style={
              Object {
                "fill": "white",
              }
            }
          />
        </div>
        <div
          className={undefined}
        >
          item_label
        </div>
      </div>,
      "className": undefined,
      "href": "#item_id",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className={undefined}
          >
            <t
              className="SVG-container"
              raw="text-file-stub"
              style={
                Object {
                  "fill": "white",
                }
              }
            />
          </div>,
          <div
            className={undefined}
          >
            item_label
          </div>,
        ],
        "className": undefined,
        "onMouseEnter": [Function],
        "onMouseLeave": [Function],
        "style": Object {
          "backgroundColor": "#323c50",
          "color": "white",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <t
              className="SVG-container"
              raw="text-file-stub"
              style={
                Object {
                  "fill": "white",
                }
              }
            />,
            "className": undefined,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "className": "SVG-container",
              "raw": "text-file-stub",
              "style": Object {
                "fill": "white",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "item_label",
            "className": undefined,
          },
          "ref": null,
          "rendered": "item_label",
          "type": "div",
        },
      ],
      "type": "div",
    },
    "type": "a",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className={undefined}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "backgroundColor": "#323c50",
              "color": "white",
            }
          }
        >
          <div
            className={undefined}
          >
            <t
              className="SVG-container"
              raw="text-file-stub"
              style={
                Object {
                  "fill": "white",
                }
              }
            />
          </div>
          <div
            className={undefined}
          >
            item_label
          </div>
        </div>,
        "className": undefined,
        "href": "#item_id",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className={undefined}
            >
              <t
                className="SVG-container"
                raw="text-file-stub"
                style={
                  Object {
                    "fill": "white",
                  }
                }
              />
            </div>,
            <div
              className={undefined}
            >
              item_label
            </div>,
          ],
          "className": undefined,
          "onMouseEnter": [Function],
          "onMouseLeave": [Function],
          "style": Object {
            "backgroundColor": "#323c50",
            "color": "white",
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <t
                className="SVG-container"
                raw="text-file-stub"
                style={
                  Object {
                    "fill": "white",
                  }
                }
              />,
              "className": undefined,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "className": "SVG-container",
                "raw": "text-file-stub",
                "style": Object {
                  "fill": "white",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "item_label",
              "className": undefined,
            },
            "ref": null,
            "rendered": "item_label",
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "a",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
